packages/gatsby/src/utils/page-data.js

```diff
@@ -22,7 +25,75 @@ const write = async ({ publicDir }, page, result) => {
   await fs.outputFile(filePath, JSON.stringify(body))
 }
 
+/**
+ * Returns the corresponding html file path for a page-data.json file
+ */
+const getPageDataHtmlPath = pageDataPath => {
+  let htmlPath = pageDataPath
+    .replace(`/index/`, `/`)
+    .replace(`./public/page-data/`, `/`)
+    .replace(`/page-data.json`, ``)
+
+  if (!htmlPath.endsWith(`.html`)) {
+    htmlPath = `${htmlPath}/`
+  }
+
+  return htmlPath
+}
+
+/**
+ * Finds a list of unused page-data.json files that are leftover from deleted pages
+ *
+ * @argument {object}
+ * {object} availablePagePaths: an array of paths to pages
+ * {object} directory: the root of the Gatsby site
+ * @returns {array}
+ */
+const getUnusedPageDataPaths = ({ availablePagePaths, directory }) => {
+  const allPageDataFiles = glob.sync(`./public/page-data/**/page-data.json`, {
+    cwd: directory,
+  })
+  const unusedPageDataFiles = allPageDataFiles.filter(pageDataPath => {
```

Comment:
This might be a problem at scale ...

Rather than explicitly looping through all pages, wouldn't it be more efficient to schedule a delete and cleanup in that step?

Consider a site with 100k pages ... This search will not be fun.